source/javascripts/site.js

```diff
@@ -1,130 +1,31 @@
+//= require _imageMasks
+
 window.onload = function(){
   var img    = document.getElementById('palmtree'),
       canvas = document.getElementById('canvas'),
-      ctx    = canvas.getContext('2d');
+      ctx    = canvas.getContext('2d')
+      slider = document.getElementById('brightness-adjustment'),
 
   canvas.width = img.width;
   canvas.height = img.height;
   ctx.drawImage(img, 0, 0, img.width, img.height);
 
   var imageData = ctx.getImageData(0, 0, img.width, img.height);
-  var data = imageData.data;
   var pixelCount = canvas.width * canvas.height;
-
-  // Color as array for rgb values
-  // TODO: generate this array from an rgba color
-  // to allow more flexibility on input here
-  var green = [80, 254, 16];
-  var red = [254, 16, 16];
-
-  // Create a gradient from which to pick colors from the duotone
-  // based on pixel luminence.
-  var gradientArray = createGradient(green, red);
-
-  var duoToneData =  convertToDuoTone(data, pixelCount, gradientArray);
+  var duoToneData = ImageMasks.convertToDuoTone(imageData.data, pixelCount);
   var imageData = new ImageData(new Uint8ClampedArray(duoToneData), canvas.width, canvas.height);
 
   ctx.putImageData(imageData, 0, 0);
-};
-
-function convertToDuoTone(imageData, pixelCount, gradientArray) {
-  var pixels = imageData;
-  var pixelArray = [];
-
-  convertImageToGrayscale(pixels);
-
-  for (var i = 0; i < pixelCount; i++) {
-    var offset, r, g, b, a;
-
-    offset = i * 4;
-    r = pixels[offset + 0];
-    g = pixels[offset + 1];
-    b = pixels[offset + 2];
-    a = pixels[offset + 3];
-
-    // Average the rgba for HSL conversion
-    // Pixel brightness math from w3 suggested formula:
-    // https://www.w3.org/TR/AERT#color-contrast
-    var average = Math.floor(((299 * r) + (587 * g) + (114 * b)) / 1000);
-    var hsl = rgbToHsl(average, average, average);
-    var luminosity = Math.max(0, Math.min(254, Math.floor((hsl[2] * 254))));
-
-    // Replace values with new values from gradient array
-    r = gradientArray[luminosity][0];
-    g = gradientArray[luminosity][1];
-    b = gradientArray[luminosity][2];
-
-    pixelArray.push(r);
-    pixelArray.push(g);
-    pixelArray.push(b);
-    pixelArray.push(a);
-  }
-
-  return pixelArray;
-};
-
 
-function convertImageToGrayscale(imageData) {
-  var r, g, b, avg;
+  slider.onchange = function() {
+    brightness = slider.value;
+    ctx.drawImage(img, 0, 0, img.width, img.height);
+    var imageData = ctx.getImageData(0, 0, img.width, img.height);
 
-  for(var pixel = 0, len = imageData.length; pixel < len; pixel += 4) {
-    r = imageData[pixel]
-    g = imageData[pixel + 1];
-    b = imageData[pixel + 2];
+    var adjustedImage = ImageMasks.adjustBrightness(imageData.data, brightness);
+    var duoToneData = ImageMasks.convertToDuoTone(adjustedImage, pixelCount);
+    var imageData = new ImageData(new Uint8ClampedArray(duoToneData), canvas.width, canvas.height);
```

Comment:
Lost a comment here in my review, but what I was referencing in the comment for the `slider.onchange` event is that we handle the data here in a way that causes a lot of unnecessary duplication. When there's a change, we re-draw the original image to get its data again before applying the image filters. This is a lot of work that gets duplicated each time! Maybe we should consider storing the original image data so that we just apply the filters to that data each time, rather than redrawing and gathering the data again? 

We also pull out the imageData.data to manipulate it before creating a new ImageData object. Is there a way we could refactor the image masks to take in a whole ImageData object so that we don't have to create a new one?